### ISSUE TITLE
Clear address and contact for billing and shipping

### ISSUE THREAD

**Issue by barredterra** [frappe/erpnext#33707]
### Problem

I want to define in advance what addresses and contacts should be used for billing and shipping in sales transactions.

In **Contact**, we have two checkboxes _Is Primary Contact_ and _Is Billing Contact_. In **Customer**, we can select a _Primary Contact_, but this is not limited to the contacts who have _Is Primary Contact_ checked. Now, when I create a **Sales Invoice**, will the system use the primary contact or the primary contact (sic)? A shipping contact cannot be specified at all.

In **Address**, we have two checkboxes _Preferred Billing Address_ and _Preferred Shipping Address_. In **Customer**, we can select a _Primary Address_, but this is not limited to addresses who have either _Preferred Billing Address_ or _Preferred Shipping Address_ checked. 

To add to that, **Address** and **Contact** are designed to be linked to multiple entities via **Dynamic Link**. What's a "Primary Contact" for one entity may be a secondary for the other. What'a a billing address for one might be a shipping address for the other. Therefore, it does not make sense to have these checkboxes here.

The naming of fields is inconsistent across transactions. In a **Sales Invoice**, the billing address goes into a field called _Customer Address_, but in a **Delivery Note** the same field is called _Billing Address Name_. This makes mapping between these documents hard.

### Proposed solution (incomplete)

Add these fields in **Customer** and all sales transactions:

- Tab: Billing
    - Link: Billing Contact
    - Link: Billing Address
- Tab: Shipping
    - Link: Shipping Contact
    - Link: Shipping Address

In new sales transactions, the addresses and contacts get fetched from the customer.

When adding an email to a transaction, the primary email address of the billing or shipping contact (depending on the transaction) is selected automatically.

_Primary Contact_ and _Primary Address_ will be removed from **Customer** and their content copied to the new fields.

**Comment by scdanieli** [comment]
@barredterra I'll write a detailed proposed solution tomorrow ...

**Comment by muchai** [comment]
Maybe an enhancement to have a setting, that disables picking say primary address on sales transactions from the customer.

In some case, where there's one billing customer, multiple shipping addresses, you want to have shipping address on sales transactions as mandatory, and blank, that way the user picks the expected shipping address.

**Comment by barredterra** [comment]
> In some case, where there's one billing customer, multiple shipping addresses, you want to have shipping address on sales transactions as mandatory, and blank, that way the user picks the expected shipping address. – @muchai 

Makes sense! I think this could be achieved in the following way, with our proposal: Leave the _Shipping Address_ in **Customer** empty and make it mandatory in the sales transaction via **Customize Form**. This way it will be empty and mandatory in the transaction, prompting the user to pick the right one.

**Comment by scdanieli** [comment]
As described by @barredterra, there are many settings to set preferred addresses and contacts and it is not very obvious what effect each setting has. But these settings are used in many, many places in ERPNext. So it's pretty tricky to change anything here. 

To solve the problem of this issue, I would be necessary to clear up the *preferred address and contact logic* in general and to set addresses and contacts explicitly at relevant places. Otherwise, an already messy part will become even messier.

That would mean getting rid of the two checkboxes **Preferred Billing Address** and **Preferred Shpping Address** in **Address**. And also of the two checkboxes **Is Primary Contact** and **Is Billing Contact** in Contact. Some existing logic could be retained or improved, but some would be lost in the process.

Since so many parts would be affected by this change, I'm not quite sure if this should be tackled or not.

**Comment by muchai** [comment]
Best to stick to your original scope. The other need can always be achieved another way, for now.

**Comment by tweetechnology** [comment]
This Can be achieved by client script and Custom Fields if i am not wrong , 
https://docs.erpnext.com/docs/v14/user/manual/en/customization/articles/fetch-the-entire-address-text-on-a-custom-field
